File: container_most_water/container.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        
        # track two indices, start from ends of the array
        i = 0
        j = len(height) - 1
        area = 0
        
        # walk inwards from ends until we meet
        # we choose which direction to walk in based on which of the 
        # two currently selected heights is smaller
        while i < j:
            area = max(area, min(height[i], height[j]) * (j - i))
            
            if height[j] > height[i]:
                i += 1
            else:
                j -= 1
            
        return area

File: container_most_water/container.java
import java.lang.Math;

class Solution {
    public int maxArea(int[] height) {
        // maintain pointers
        int right = height.length - 1;
        int left  = 0;
        
        int champion = 0;
        int area = 0;        
        
        // move pointers inwards until they meet
        while (left < right) {
            area = Math.min(height[left], height[right]) * (right - left);
            if (area > champion) champion = area;
            
            // move the pointer with smaller height, inwards
            if (height[left] > height[right]) right--;
            else left++;
        }
        
        return champion;    
    }
}
